[Geographiclib] update to v2.0

cffk opened the pull request:
**Update to GeographicLib 2.0**

- #### What does your PR fix?
  Reorganization and simplification of code base.  A couple of minor bug fixes.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  Should work on all platforms with C++11 compilers

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

cffk committed:
Update for geographiclib 2.0

cffk committed:
[geographiclib] Update to version 2.0

cffk commented:
> In the ports that use the new function, you have to add the corresponding dependencies:
> 
> ```json
> {
>   "name": "vcpkg-cmake",
>   "host": true
> },
> {
>   "name": "vcpkg-cmake-config",
>   "host": true
> }
> ```

I can't figure out where to put these dependencies.

cffk committed:
Add license to ports/geographiclib/vcpkg.json

cffk commented:
> _This is a new experimental fast check for PR issues. Please let us know if this bot is helpful!_
> PRs must add only one version and must not modify any published versions
> 
> You have modified or added at least one portfile where deprecated functions are used.
> 
> If you feel able to do so, please consider migrating them to the new functions: `vcpkg_install_cmake` -> `vcpkg_cmake_install` (from port `vcpkg-cmake`) `vcpkg_build_cmake` -> `vcpkg_cmake_build` (from port `vcpkg-cmake`) `vcpkg_configure_cmake` -> `vcpkg_cmake_configure` (Please remove the option `PREFER_NINJA`) (from port `vcpkg-cmake`) `vcpkg_fixup_cmake_targets` -> `vcpkg_cmake_config_fixup` (from port `vcpkg-cmake-config`)
> 
> In the ports that use the new function, you have to add the corresponding dependencies:
> 
> ```json
> {
>   "name": "vcpkg-cmake",
>   "host": true
> },
> {
>   "name": "vcpkg-cmake-config",
>   "host": true
> }
> ```
> 
> The following files are affected:
> 
>     * `ports/geographiclib/portfile.cmake`

I don't know to put the dependency information.  Please make the diagnostic less cryptic.

cffk committed:
Update port-version to 1

cffk committed:
Update version database

BillyONeal commented on ports/geographiclib/portfile.cmake:
```diff
@@ -45,5 +50,3 @@ file (INSTALL ${SOURCE_PATH}/LICENSE.txt
 configure_file (${CMAKE_CURRENT_LIST_DIR}/usage
   ${CURRENT_PACKAGES_DIR}/share/${PORT}/usage @ONLY)
-
-vcpkg_fixup_pkgconfig()
```
Why is this gone?

cffk replied:
It didn't look like it was doing anything.  If you tell me it's needed I'm happy to put it back.

Adela0814 replied:
[this function](https://github.com/microsoft/vcpkg/blob/bec883fef498206f05b20bc0a4ba241637657dba/docs/maintainers/vcpkg_fixup_pkgconfig.md) will fix common paths in `*.pc` files and make everything relative to `$(prefix)`.

cffk replied:
I restored this.

Adela0814 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
+  set (TOOL_OPTION "-DBINDIR=")
 endif ()
 
 vcpkg_configure_cmake (
```
This is a deprecated function. Please change it to `vcpkg_cmake_configure`

cffk replied:
Done

Adela0814 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
     PREFER_NINJA # Disable this option if project cannot be built with Ninja
 )
 
 vcpkg_install_cmake ()
```
vcpkg_cmake_install

cffk replied:
Done

Adela0814 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
 
 vcpkg_install_cmake ()
-vcpkg_fixup_cmake_targets (CONFIG_PATH share/geographiclib)
+vcpkg_fixup_cmake_targets ()
```
vcpkg_cmake_config_fixup, refer to [this](https://github.com/microsoft/vcpkg/blob/f44d4c169f1b81a6188b330dbd837881ffb9ef2b/docs/maintainers/vcpkg_fixup_cmake_targets.md)

cffk replied:
Done

Adela0814 commented:
> > _This is a new experimental fast check for PR issues. Please let us know if this bot is helpful!_
> > PRs must add only one version and must not modify any published versions
> > You have modified or added at least one portfile where deprecated functions are used.
> > If you feel able to do so, please consider migrating them to the new functions: `vcpkg_install_cmake` -> `vcpkg_cmake_install` (from port `vcpkg-cmake`) `vcpkg_build_cmake` -> `vcpkg_cmake_build` (from port `vcpkg-cmake`) `vcpkg_configure_cmake` -> `vcpkg_cmake_configure` (Please remove the option `PREFER_NINJA`) (from port `vcpkg-cmake`) `vcpkg_fixup_cmake_targets` -> `vcpkg_cmake_config_fixup` (from port `vcpkg-cmake-config`)
> > In the ports that use the new function, you have to add the corresponding dependencies:
> > ```json
> > {
> >   "name": "vcpkg-cmake",
> >   "host": true
> > },
> > {
> >   "name": "vcpkg-cmake-config",
> >   "host": true
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The following files are affected:
> > ```
> > * `ports/geographiclib/portfile.cmake`
> > ```
> 
> I don't know to put the dependency information. Please make the diagnostic less cryptic.

you can add dependency information to `ports/geographiclib/vcpkg.json`

Adela0814 commented on ports/geographiclib/vcpkg.json:
```diff
@@ -1,13 +1,13 @@
-  "version": "1.52",
-  "port-version": 2,
+  "version": "2.0",
+  "port-version": 1,
```
the `port-version` should be removed if you changed the `version`, the default value of `port-version` is `0`

cffk replied:
I see that, in order to stick with port-version = 0, I have to remove the previous version information before rerunning x-add-version.  I've done this.

Adela0814 commented on ports/geographiclib/vcpkg.json:
```diff
@@ -1,13 +1,13 @@
   "features": {
     "tools": {
-      "description": "The GeographicLib tools",
-      "supports": "!uwp"
```
why you remove supports? is `uwp `supported now?

cffk replied:
I've added this back.  (Not sure why I removed it in the first place.)

cffk committed:
Switch to vcpkg_cmake_config_fixup etc.

Switch to
  vcpkg_cmake_configure
  vcpkg_cmake_install,
  vcpkg_cmake_config_fixup

Revert to using vcpkg_fixup_pkgconfig
Revert to port-version = 0
Revert to uwp exclusion to tools

cffk committed:
run x-add-version

cffk commented:
> you can add dependency information to `ports/geographiclib/vcpkg.json`

Ah, the original diagnostic didn't make it clear that these were dependencies (and I don't live and breathe json files).  I've add these now.

dg0yt commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
+  REF distrib-C++
+  FILENAME "GeographicLib-2.0.tar.gz"
+  SHA512 7cf67174a64082372cdd249a64460e9f61c582aaf3d2a31e4e69d811f265e078ba62f945e9f1f44be6c58de4c20d0359dd46e0fd262ffac229df0ba2c6adc848
+  )
```
I would prefer if the formatting would remain as common in vcpkg. (Intent with four space characters.)

cffk replied:
Done.  (I'm use the narrow spacing because I work with an 80-column editor.)

dg0yt commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
+  "-DDOCDIR="
+  "-DEXAMPLEDIR="
+  "-DMANDIR="
+  "-DSBINDIR=")
```
If it is common, why not list this directly with `vcpkg_cmake_configure`?

cffk replied:
Done

Adela0814 replied:
You should delete the previous if you moved the options to `vcpkg_cmake_configure`. You can also just add the `COMMON_OPTIONS` to `vcpkg_cmake_configure` as a variable. 
But I'm confused why you added these options, is it necessary?

Adela0814 replied:
Option does not need double quote, nor does it need to add an equal sign if it does not set a value.

dg0yt commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,34 +1,39 @@
-        "-DGEOGRAPHICLIB_LIB_TYPE=${LIB_TYPE}"
-        ${FEATURE_OPTIONS}
+    OPTIONS ${LIB_OPTION} ${COMMON_OPTIONS} ${TOOL_OPTION}
     PREFER_NINJA # Disable this option if project cannot be built with Ninja
```
`PREFER_NINJA` is obsolete now, with `vcpkg_cmake_configure`.

cffk replied:
Removed.

dg0yt commented on ports/geographiclib/portfile.cmake:
```diff
@@ -46,4 +51,4 @@ file (INSTALL ${SOURCE_PATH}/LICENSE.txt
   ${CURRENT_PACKAGES_DIR}/share/${PORT}/usage @ONLY)
 
-vcpkg_fixup_pkgconfig()
+vcpkg_fixup_pkgconfig ()
```
No space before `()`.

cffk replied:
Done (consistently).

dg0yt reviewed (commented):
I would prefer if the portfile would follow common style in vcpkg. Apart from some local comments, I would suggest to place the installation of `usage` and `copyright` at the end of the file.

cffk committed:
Formatting changes requested by dg0yt

cffk committed:
Update version info

cffk committed:
More cleanup

cffk committed:
Update version info

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
-        "-DGEOGRAPHICLIB_LIB_TYPE=${LIB_TYPE}"
-        ${FEATURE_OPTIONS}
-    PREFER_NINJA # Disable this option if project cannot be built with Ninja
+    OPTIONS ${LIB_OPTION} ${TOOL_OPTION}
+    "-DCMAKEDIR=share/${PORT}"
+    -DPKGDIR=
+    -DDOCDIR=
+    -DEXAMPLEDIR=
+    -DMANDIR=
+    -DSBINDIR=
```
```suggestion
    OPTIONS
        "-DBINDIR=tools/${PORT}"
        "-DCMAKEDIR=share/${PORT}"
        "-DDOCDIR=share/${PORT}/doc"
        "-DEXAMPLEDIR=share/${PORT}/doc"
        "-DMANDIR=share/${PORT}/man"
        "-DSBINDIR=share/${PORT}/tools"
```

`LIB_OPTION` is not defined.
`BINDIR` should always be set to _tools/${PORT}_.
We should keep `PKGDIR` default value and call `vcpkg_fixup_pkgconfig()` after call `vcpkg_cmake_install`.

cffk replied:
I defined LIB_OPTION (a mistake on my part that it wasn't set).
I keep PKGDIR set so the pc file is installed.
I didn't set BINDIR to tools/${PORT} because in a previous submission I was told that tools should not be installed unless the tools option is given.  Also the original port (not by me) skipped installing the man pages, doc, etc.  Where should I look for the vcpkg standards on this?

JackBoosY replied:
Tools should use option `TOOLS` to decide whether to enable or not, not based on this path.

dg0yt replied:
> Tools should use option `TOOLS` to decide whether to enable or not, not based on this path.

I would have chosen a different wording. But I do share the concern:
- I would interprete empty paths as a request to use the default - similar to unset variables.
- I would prefer a clear variable names to control options (`BUILD_TOOLS`, `ENABLE_TOOLS` or `WITH_TOOLS`).

Of course, it is upstream to decide whether to change it in a future release. For 2.0, vcpkg has to use the interface which is provided. So this part of the PR is okay: Setting an empty path means turning the feature off. Maybe this should be written into a comment, because it is unusual enough to surprise future readers.

cffk replied:
In order to deal with the varying needs of multiple packagers, I decided that specifying an install path for each component with an empty path meaning "don't install this component" was the most flexible approach. This let me dispense with a 142-line patch file and replace it with half a dozen install directories set to null.

The logic in CMakeLists does not distinguish `-DBINDIR=OFF` and `-DBINDIR=`, so I've changed the portfile to use `OFF` to mean don't install.  Maybe that will reduce the confusion.  I've also added a comment to explain the behavior

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
-else ()
-  set (LIB_TYPE "STATIC")
-endif ()
+if(TOOLS)
+    set(TOOL_OPTION "-DBINDIR=tools/${PORT}")
+else()
+    set(TOOL_OPTION -DBINDIR=)
+endif()
```
```suggestion
```

cffk replied:
In a previous submission (for version 1.52, I think) I was told that tools should not be installed unless the tools option is given.  Has this guidance changed?  Where do I have the current standards?

JackBoosY replied:
This modification can make the code more concise: 
- if we do not install the tool, this variable is not used
- if we install the tool, the value of this variable is used. 

So whether the tool is enabled or not, we should set this value.

dg0yt replied:
geographiclib's interpretation of `BINDIR` is confusing. But this part of the PR is okay IMO.

cffk replied:
GeographicLib's usage of these install directory options is now documented.  See above.

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
+
+vcpkg_check_features(
+    OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES "tools" TOOLS
```
```suggestion
    FEATURES
        "tools" TOOLS
```

cffk replied:
done

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
 
-vcpkg_configure_cmake (
+vcpkg_cmake_configure(
     SOURCE_PATH ${SOURCE_PATH}
```
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```
Please add double quotes to all the absolute paths.

cffk replied:
done

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
-# Install usage
-configure_file (${CMAKE_CURRENT_LIST_DIR}/usage
-  ${CURRENT_PACKAGES_DIR}/share/${PORT}/usage @ONLY)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/tools)
```
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/tools")
```

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
-  vcpkg_copy_tool_dependencies (${CURRENT_PACKAGES_DIR}/tools/${PORT})
-endif ()
+if(tools IN_LIST FEATURES)
+    vcpkg_copy_tool_dependencies(${CURRENT_PACKAGES_DIR}/tools/${PORT})
```
```suggestion
    vcpkg_copy_tool_dependencies("${CURRENT_PACKAGES_DIR}/tools/${PORT}")
```

cffk replied:
done

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
 
 vcpkg_fixup_pkgconfig()
+
+file(INSTALL ${SOURCE_PATH}/LICENSE.txt
+    DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
```
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE.txt"
    DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
```

cffk replied:
done

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
+    RENAME copyright)
+
+# Install usage
+configure_file(${CMAKE_CURRENT_LIST_DIR}/usage
+    ${CURRENT_PACKAGES_DIR}/share/${PORT}/usage @ONLY)
```
```suggestion
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage"
    "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" @ONLY)
```

cffk replied:
done

JackBoosY commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,51 @@
-vcpkg_install_cmake ()
-vcpkg_fixup_cmake_targets (CONFIG_PATH share/geographiclib)
-vcpkg_copy_pdbs ()
+vcpkg_cmake_install()
```
```suggestion
vcpkg_cmake_install()

vcpkg_fixup_pkgconfig()
```

cffk replied:
Well I was already calling `vcpkg_fixup_pkgconfig`.  I moved the position of the call.

cffk committed:
Define LIB_OPTION, fix pkg install, etc.

cffk committed:
Update version info

cffk committed:
Fix pkgconfig call

cffk committed:
Update version info

cffk committed:
Use OFF instead of empty string for don't install

cffk committed:
Update version info

Thomas1664 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,61 @@
+    set(TOOL_OPTION -DBINDIR=OFF)
+endif()
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    set(LIB_OPTION -DBUILD_SHARED_LIBS=ON)
+else()
+    set(LIB_OPTION -DBUILD_SHARED_LIBS=OFF)
+endif()
```
```suggestion
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" LIB_OPTION)
```

dg0yt replied:
We don't need `LIB_OPTION` at all. `BUILD_SHARED_LIBS` is a standard CMake variable, already handled by `vcpkg_cmake_configure`...

Thomas1664 replied:
Yes, @dg0yt is absolutely right.

cffk replied:
Removed `LIB_OPTION` stuff.

Thomas1664 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,61 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS ${LIB_OPTION} ${TOOL_OPTION}
+    "-DCMAKEDIR=share/${PORT}"
+    -DDOCDIR=OFF
+    -DEXAMPLEDIR=OFF
+    -DMANDIR=OFF
+    -DSBINDIR=OFF
```
```suggestion
    OPTIONS
        ${TOOL_OPTION}
        -DBUILD_SHARED_LIBS=${LIB_OPTION}
        "-DCMAKEDIR=share/${PORT}"
        -DDOCDIR=OFF
        -DEXAMPLEDIR=OFF
        -DMANDIR=OFF
        -DSBINDIR=OFF
```

Thomas1664 commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,61 @@
 
-vcpkg_fixup_pkgconfig()
+# Install usage
+configure_file(${CMAKE_CURRENT_LIST_DIR}/usage
```
```suggestion
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage"
```

cffk replied:
Done

Thomas1664 reviewed (commented):
Minor nits, not a must to implement

cffk committed:
Remove unneeded LIB_OPTION stuff.

cffk committed:
Update version info

cffk committed:
Add more precautionary quotes

cffk committed:
Update version info

dg0yt commented on ports/geographiclib/portfile.cmake:
```diff
@@ -1,49 +1,55 @@
-vcpkg_fixup_pkgconfig()
+# Install usage
+configure_file("${CMAKE_CURRENT_LIST_DIR}/usage"
+    "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" @ONLY)
```
You already did a great job handling all the review request. This is only a nice-to-have: Rewrite `usage` so that it can be installed directly. The port name is fixed. No need to print the triplet. (This simplifications was already done in many other ports.) Cf. port `proj` for an example.

cffk replied:
Let's wait on this for the next version.

BillyONeal commented:
Thanks for the update!
